### PR TITLE
Fix cilium-health node updates

### PR DESCRIPTION
**Pull request by joestringer**
Fix the handling of updates for existing nodes in `cilium-health`. The first two commits are bugfixes; the last two are to unit-test the functionality. Read commit-by-commit for further details.

Broadly this PR removes the previous "mark-and-sweep" GC logic for nodes that previously existed in the `cilium-health` prober, because since PR #8300 we have had the exact context of added/removed nodes, meaning that we should not need to sweep the entire list.

There were two subtle bugs in the previous implementation:
* When a node is updated, `cilium-health` would continue to probe IPs that no longer correspond to the node
* Removal of a node would not disable TCP probing for all IPs corresponding to the node, it would only stop probing the primary IP of the node

Functionally when a node is updated (eg the health endpoint IP changes), the Cilium nodes API will propagate notifications that include the node in the `removed` nodes list as well as the `added` nodes list, and in this PR this will cause a reset of the node's cached prober results. It's not clear to me right now that there's any advantage to ensuring that the internally cached node representation inside the `cilium-health` prober would benefit from actually retaining the existing entry in the nodes map if the individual IP doesn't change; it's simpler to just handle the update in this scenario by actually deleting and adding the node. The following probe cycle will subsequently update the results for the IPs corresponding to the changed node. Strictly speaking this means there may be intermittent periods when nodes are updated where the `cilium-health` instance won't have prober results for the updated nodes. 

Related: #9103 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9491)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Review by gandro — APPROVED** [review]
Nice!

File: pkg/ip/addr.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2019 Authors of Cilium
```
Comment (aanm) [inline pkg/ip/addr.go]: This should not be required or at least the full file. See [1]

File: pkg/health/server/prober_test.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+					Enabled: false,
+				},
+			},
+		},
+	}, net.ParseIP(nodeIP), net.ParseIP(healthIP)
+}
+
+func sortNodes(nodes map[string][]*net.IPAddr) map[string][]*net.IPAddr {
+	for _, slice := range nodes {
+		sort.Sort(ip.AddrsByMask(slice))
```
Comment (aanm) [inline pkg/health/server/prober_test.go]: [1]
```golang
	sort.Slice(slice, func(i, j int) bool {
		iLength := len(slice[i].IP)
		jLength := len(slice[j].IP)
		if iLength == jLength {
			return bytes.Compare(slice[i].IP, slice[j].IP) < 0
		}
		return iLength < jLength
	})
```

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by joestringer** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**